Add methods to check for integer constants one and minus one

CarolEidt opened the pull request:
These are primarily motivated by simplifying the RyuJIT tutorial,
but also make the existing code a bit cleaner.

CarolEidt commented:
@dotnet/jit-contrib PTAL

sivarv commented on src/jit/gentree.h:
```diff
@@ -4097,6 +4101,30 @@ inline bool GenTree::IsZero()
+
+    if ((gtOper == GT_CNS_LNG) && (gtIntConCommon.LngValue() == 1))
+        return true;
+
```
It would be confusing if this method is defined on GenTree and doesn't handle floating-point constants.  I would suggest to do one of the following:

1) Rename the method to IsIntegralOne() to be indicative of its purpose
2) Add handling for floating point constants as well
e.g.
`if ((gtOper == GT_CNS_DBL) && (gtDblCon.gtDconVal == 1.0))`

CarolEidt replied:
That's why I called it "IsIOne" instead of IsOne. I can go for IsIntegralOne instead.

sivarv replied:
Sorry, I have missed "I" in the name :)

sivarv commented on src/jit/gentree.h:
```diff
@@ -4097,6 +4101,30 @@ inline bool GenTree::IsZero()
+}
+
+// Note that, unlike IsZero(), this only returns true only for integral types.
+inline bool GenTree::IsIMinusOne()
```
As I said above, we should either rename this to IsIntegralMinusOne() or handle floating-point constants as well.

sivarv commented on src/jit/gentree.h:
```diff
@@ -4097,6 +4101,30 @@ inline bool GenTree::IsZero()
     return false;
 }
 
+// Note that, unlike IsZero(), this only returns true only for integral types.
```
(Minor) word "only" got repeated twice.

".. this only trues true for integral types"

Similar correction is needed on line 4116

sivarv commented on src/jit/morph.cpp:
```diff
@@ -7722,7 +7722,7 @@ GenTreePtr          Compiler::fgMorphCall(GenTreeCall* call)
         GenTreePtr value = gtArgEntryByArgNum(call, 2)->node;
 
-        if (value->OperGet() == GT_CNS_INT && value->AsIntConCommon()->IconValue() == 0)
+        if (value->IsZero())
```
This will be true if value is a floating point const zero.  We need to check that it is an int const only here?

IMO, the right abstractions is

IsIntZero()
IsLngZero() - defined only for 32-bit targets
IsFloatZero() 
IsIntegralZero() - returns true if either of IsIntZero() or IsLngZero() is true
IsZero() - returns IsIntegralZero() || isFloatZero()

IsIntOne() - GT_CNS_INT and value is 1
IsLngOne() - defined only on 32-bit targets
IsIntegralOne() - returns IsIntOne() || IsLngOne()

IsIntMinusOne()
IsLngMinusOne()
IsIntegralMinusOne()

Now this check could be written as

`if (value->IsIntZero())`

CarolEidt replied:
It is foolish to have abstractions that are unused, or rarely used. In this particular context, we are looking at a particular helper call, and we know the arg type is int. We don't need to check.
I have selected the abstractions that are useful.
Please note that you may not like the name, but what I have implemented is IsIntegralZero() - and there is no need to add extra qualification when we already know the type.

pgavlin replied:
It might be nice to have an `assert(value->OperGet() == GT_CNS_INT)` for the benefit of future readers.

briansull replied:
I think that the easiest way forward would be just to add just these four methods:
IsIntOne
IsIntegralOne
IsIntMinusOne
IsIntegralMinusOne

Is that a satisfactory way forward?

sivarv commented on src/jit/morph.cpp:
```diff
@@ -9190,7 +9190,7 @@ GenTree* Compiler::fgMorphRecognizeBoxNullable(GenTree* compare)
     }
 
-    if (opCns->gtIntConCommon.IconValue() != 0)
+    if (!opCns->IsZero())
```
In principle this is checking a more conservative condition than required here. The reason it will not be a correctness issue is that if opCns is anything other than an GT_CNS_INT, execution will not reach here.  But the right condition to check here is

`If (!opCns->IsIntZero())`

CarolEidt replied:
I disagree that this is a bad thing - because the conservatism is minor, while adding many flavors of Zero just pollutes the API.

sivarv commented on src/jit/morph.cpp:
```diff
@@ -9550,8 +9550,7 @@ GenTreePtr          Compiler::fgMorphSmpOp(GenTreePtr tree, MorphAddrContext* ma
             
-            noway_assert( (effOp1->gtOper == GT_CNS_INT) &&
-                         ((effOp1->gtIntCon.gtIconVal == 0) || (effOp1->gtIntCon.gtIconVal == 1)) );
+            noway_assert(effOp1->IsZero() || effOp1->IsIOne());
```
Here too the condition has become more conservative and allows float/long zero and long one as permissible values.  The right condition here is

effOp1->IsIntZero() || effOp1->IsIntOne()

CarolEidt replied:
I think that you actually mean that the condition is more liberal. However, by definition op1 will be of TYP_INT. So, although this adds (minor) extra liberality to the check, it keeps the code simple.

sivarv replied:
In this specific instance liberal/conservative condition might be fine because we are checking in the context of a specific helper.  But in general I am against doing conservative checks where it is not needed as it could lead to hard to diagnose bugs.  Besides, it is easy to write an exact condition here.

sivarv replied:
Besides I don't feel comfortable with liberal checks and depending on the context for those checks to be correct for two reasons: 1) it could introduce hard to diagnose bugs 2) for a person trying to read the code it gives wrong understanding of the code potentially unless one is very careful.

CarolEidt replied:
I think we are at an impasse here. I strongly oppose cluttering the API surface with excess variants without just cause. I don't think these contexts are at all difficult to understand, nor would it be easy to get confused about these.

sivarv replied:
To make myself clear:
I am not opposed to adding new abstraction IsIOne() and IsIMinusOne() and using it only those places where we don't make the condition liberal and depend on context around it for correctness.

pgavlin replied:
> I don't think these contexts are at all difficult to understand

They may be for new readers. I personally would feel more comfortable if the assumptions being made at these various checks were covered by assertions.

sivarv commented on src/jit/morph.cpp:
```diff
@@ -10831,10 +10829,7 @@ GenTreePtr          Compiler::fgMorphSmpOp(GenTreePtr tree, MorphAddrContext* ma
-                        goto SKIP;
-
-                    if (andOp->gtOp.gtOp2->gtIntCon.gtIconVal != 1)
+                    if (!andOp->IsIOne())
```
This would allow long const one, previous which wasn't allowed.

sivarv replied:
This change introduces a correctness bug as the previous condition was on andop->gtOp.gtOp2 (i.e. second operand of andOp whereas this is on andOp itself.

sivarv commented on src/jit/morph.cpp:
```diff
@@ -10855,10 +10850,7 @@ GenTreePtr          Compiler::fgMorphSmpOp(GenTreePtr tree, MorphAddrContext* ma
-                        goto SKIP;
-
-                    if (andOp->gtOp.gtOp2->gtIntConCommon.LngValue() != 1)
+                    if (!andOp->IsIOne())
```
Here too we are checking whether andOp is cns one instead what we should be checking is second operand of andOp

sivarv replied:
GT_CNS_NATIVELONG is defined as GT_CNS_INT on 64-bit and GT_NS_LNG on32-bit targets.  Once again the above check is more liberal than required and not giving the right picture to a person who reads this code.

sivarv commented on src/jit/morph.cpp:
```diff
@@ -10969,7 +10961,7 @@ GenTreePtr          Compiler::fgMorphSmpOp(GenTreePtr tree, MorphAddrContext* ma
                 cns2 = op2;
                 /* Check for "expr relop 1" */
-                if (cns2->gtIntCon.gtIconVal == +1)
+                if (cns2->IsIOne())
```
This condition allows long const one and not a correctness issue here because of the context around this condition.  We should avoid these.   Plus is it is unnecessarily checking whether cns2 is GT_CNS_INT, though it is already checked above.

pgavlin commented on src/jit/gentree.h:
```diff
@@ -4097,6 +4101,30 @@ inline bool GenTree::IsZero()
+    if ((gtOper == GT_CNS_LNG) && (gtIntConCommon.LngValue() == 1))
+        return true;
+
+    return false;
```
Nit: this whole function could be reduced to `return ((gtOper == GT_CNS_INT) && (gtIntConCommon.IconValue() == 1) || ((gtOper == GT_CNS_LNG) && (gtIntConCommon.LngValue() == 1));` (similarly for `IsIMinusOne`). If you feel that the current style is more readable, that's fine--I don't have a particularly strong opinion here.

sivarv commented on src/jit/morph.cpp:
```diff
@@ -12453,22 +12445,12 @@ GenTree* Compiler::fgMorphSmpOpOptional(GenTreeOp* tree)
                     /* Special case: "x |= -1" and "x &= 0" */
-                    if  (cmop == GT_AND || cmop == GT_OR)
+                    if (((cmop == GT_AND) && op2->gtOp.gtOp2->IsZero()) ||
+                        ((cmop == GT_OR)  && op2->gtOp.gtOp2->IsIMinusOne()))
```
Another instance of liberal condition and making the assumption that we never generate certain IR nodes?

sivarv commented on src/jit/morph.cpp:
```diff
@@ -12739,20 +12715,13 @@ GenTree* Compiler::fgMorphSmpOpOptional(GenTreeOp* tree)
-                DEBUG_DESTROY_NODE(op2);
-            }
-            else if ((op2->gtOper == GT_CNS_LNG) && (op2->gtIntConCommon.LngValue() == -1))
+            if (op2->IsIMinusOne())
```
This is a perfect example of a case where the new abstraction fits perfectly without liberalizing the condition nor depending on the surrounding execution context.  I like these kind of changes.

sivarv commented on src/jit/morph.cpp:
```diff
@@ -15130,11 +15099,8 @@ void Compiler::fgCheckQmarkAllowedForm(GenTree* tree)
-    assert(colon->gtOp.gtOp2->gtOper == GT_CNS_INT);
-    assert(colon->gtOp.gtOp2->AsIntCon()->IconValue() == 1);
+    assert(colon->gtOp.gtOp1->IsZero());
+    assert(colon->gtOp.gtOp2->IsIOne());
```
Another instance of liberal check and making an assumption that certain kind/type of IR nodes never gets created.  What if this assumption gets violated and do we know that some place in JIT will catch it rather than leading to SBCG?

CarolEidt replied:
There is no way that liberalizing this check will lead to SBCG. We have many checks that operands are compatible - so we couldn't have a float 0 and an integral 1. Further, this is on legacy backend only, and we currently only allow colon to have TYP_I_IMPL operands. The point of these asserts are to ensure that the operands are the constants zero and one, NOT to ensure that the operand types have been kept consistent. So, this gets at the real point of the asserts. I actually believe quite strongly that an excess of asserts impedes rather than improves the maintainability of the code.

sivarv commented on src/jit/optimizer.cpp:
```diff
@@ -993,8 +993,7 @@ bool Compiler::optIsLoopTestEvalIntoTemp(GenTreePtr testStmt, GenTreePtr* newTes
         (opr1->OperGet() == GT_LCL_VAR) &&
-        (opr2->OperGet() == GT_CNS_INT) && 
-        (opr2->gtIntCon.gtIconVal == 0))
+        opr2->IsZero())
```
This would permit long and floating const zero, which wasn't permitted earlier.  Can it lead to correctness issues due to liberal condition being checked here?

CarolEidt replied:
I have re-added the check for GT_CNS_INT.

sivarv commented on src/jit/optimizer.cpp:
```diff
@@ -7845,12 +7844,9 @@ GenTree *           Compiler::optIsBoolCond(GenTree *   condBranch,
         isBool = true;
     }
-    else if (opr1->gtOper == GT_CNS_INT)
+    else if (opr1->IsZero() || opr1->IsIOne())
```
This would permit long and floating const zero, which wasn't permitted earlier. Can it lead to correctness issues due to liberal condition being checked here?

CarolEidt replied:
I have looked carefully at the context, and I believe that it is correct.

CarolEidt commented:
Changes based on feedback, including Brian's suggested change, and fixed the correctness issue.

mikedn commented:
Kind of late to the party but I wonder if we really need all these functions. Why `IsIntOne()` and `IsIntMinusOne()` instead of just `IsIntCon(size_t value)`? 

And why all this back and forth about `IsZero` and floating point when `IsZero` doesn't appear to ever be used to check for fp 0? Anyway using `IsZero` for floating point may be problematic due to fp 0 being signed. I'd rather remove the fp support from `IsZero` and change its name to `IsIntZero`.

CarolEidt commented:
@mikedn - there are a couple of places where `IsZero()` is used in a fp-specific context, and one where it is used in an int-or-fp context. We could add `IsFPZero()` and `IsIntegralConst(int value)` instead of all these. We generally don't abbreviate constant with con - it's usually cns (which IMO is just too terse) or const. Also, I would avoid `size_t` for crossgen reasons (and we will presumably only be checking values that fit in an `int`).
@dotnet/jitcontrib - thoughts? It would be good to get some additional feedback before iterating again. I've looked at the contexts for Int versus Integral (int or long) and I don't really see a lot of value in having both flavors.

pgavlin commented:
I think I prefer `IsIntegralConstant(int64_t value)` and `IsFPZero()`, perhaps with the addition of `IsDoubleConstant(double value)` and `IsSingleConstant(float value)`.

FWIW, here are some of LLVM's methods for inquiring about constants: http://llvm.org/docs/doxygen/html/classllvm_1_1Constant.html. They use something close to what Carol originally proposed.

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -1447,13 +1447,13 @@ void CodeGen::genCodeForBinary(GenTree* treeNode)
         && !treeNode->gtOverflowEx())
     {
-        if (src->gtIntConCommon.IconValue() == 1)
+        if (src->IsIntegralOne())
```
The if-check preceding this one has already established it is GT_CNS_INT.  But IsIntegralOne() is once again checking whether gtOper is GT_CNS_INT or GT_CNS_LNG, which is unnecessary.  To be precise, above and on line 1456 we should use IsIntOne() or IsIntMinusOne().  Even then we incur additional gtOper check inside IsIntOne() and IsIntMinusOne(). I see this excessive check in many places due to the replacement of original condition by IsIntOne/IsIntegralOne/IsIntMinusOne/IsIntegralMinusOne().

The question I have is, Is native compiler clever enough to optimize out additional gtOper check done by IsIntOne()/IsIntMinuOne() or is it best to avoid it in the first place? Note that this we need to see not just above but in all the places where usage of IsIntXXX()/IsIntegralXXX() has resulted in an additional gtOper check.

I am thinking whether these methods be defined on GenTreeIntConCommon and GenTreeDblCon instead of on GenTree?  Another advantage is that we can drop the "type" from these method names when defined on GenTreeIntCon, GenTreeLngCon and GenTreeDblCon

e.g
/ / actual methods
bool GenTreeIntCon::IsOne()
bool GenTreeIntCon::IsZero()
bool GenTreeIntCon::IsMinusOne()
bool GenTreeLngCon::IsOne()
bool GenTreeLngCon::IsZero()
bool GenTreeLngCon::IsMinusOne()
bool GenTreeDblCon::IsZero()

In case if we want to have similar methods defined on GenTree, we can do so and such a method implementation based on gtOper will cast GenTree node to specific *Con node and call the corresponding method on it.  For example GenTree:IsZero() could be implemented as follows assuming GenTreeIntCon::IsZero(), GenTreeLngCon::IsZero and GenTreeDblCon::IsZero() are available

```
bool GenTree::IsZero() 
{
   if (OperGet() == GT_CNS_INT)
   {
       this->AsIntCon()->IsZero();
   }
  else if (OperGet() == GT_CNS_LNG)
   {
        this->AsLngCon()->IsZero();
   }
  else if (OperGet() == GT_CNS_DBL) 
  {
       this->AsDblCon()->IsZero();
  }
  return false;
}

```
With the above suggestion, if-checks on lines 1450 and 1456 would become as follows

```
if (src->AsIntCon()->IsOne()) 
if (src->AsIntCon()->IsMinusOne())
```
Note: Ideally if we used virtual methods it would have allowed us to write
`if (src->IsOne()) or if (src->IsMinusOne())`.  Since we don't use virtual methods in Gentree nodes for efficiency, either 1) we have to check gtOper of the node, cast the node statically to right type and call the method IsOne()/IsMinusOne() on precise Gentree type node (**OR**) 2) GenTree::IsOne()/IsMinusOne() implementation should check gtOper and return the right result.   If there are many places where we already check gtOper and just further need to check whether const value is 0, 1 or -1, option 1 would avoid unnecessary additional gtOper check.  Option 2 would be useful in places where we need to do both gtOper check and const value check together.

mikedn commented:
> Also, I would avoid  size_t  for crossgen reasons (and we will presumably only be checking values that fit in an  int ).

Sorry, I wanted to say `ssize_t`, not `size_t`. I'm not sure what crossgen has to do with it, my suggestion was based on the type of `GenTreeIntCon::gtIconVal` which is also `ssize_t`.

> there are a couple of places where  IsZero()  is used in a fp-specific context, and one where it is used in an int-or-fp context

Yep, after I posted I had more time to look at the various uses of `IsZero` and in a couple of places (codegen trying to generate a `xor`) it's used for both integer and fp. However, checking how `IsZero` is used was an interesting exercise. While it can be deduced from context what type of zero is expected I think that the code would be easier to read if the type was specified in the function name.

I'll add that in `Lowering::TreeNodeInfoInitSIMD` there's some code that could benefit from `IsIntegralConstant(ssize_t)`:

```C++
(simdTree->gtSIMDBaseType == TYP_INT && op1->IsCnsIntOrI() && op1->AsIntConCommon()->IconValue() == 0xffffffff) ||
(simdTree->gtSIMDBaseType == TYP_LONG && op1->IsCnsIntOrI() && op1->AsIntConCommon()->IconValue() == 0xffffffffffffffffLL)
```
(hmm, I suspect that this code won't work on x86 due to mismatch between ssize_t and 0xffffffffffffffffLL...)

CarolEidt commented:
@mikedn - thanks for the `IsIntegralConst` suggestion - I've incorporated that, and changed all the existing `IsZero` uses to use `IsIntegralConst(0)` or `IsFPZero()` or both, depending on context.
I believe this is the best approach.
Also, @mikedn thanks for identifying the SIMD issue with checking for all ones. Since the constants are kept as signed values, the right check there is really IsIntegralConst(-1).
I feel pretty strongly that we should depend on the native compiler to optimize cases where we check for an opcode twice, in the interest of simpler code.

CarolEidt commented:
@dotnet-bot test Windows_NT x64 Debug Build and Test

mikedn commented on src/jit/optimizer.cpp:
```diff
@@ -7845,12 +7844,9 @@ GenTree *           Compiler::optIsBoolCond(GenTree *   condBranch,
         isBool = true;
     }
-    else if (opr1->gtOper == GT_CNS_INT)
+    else if (opr1->IsIntegralConst(0) || opr1->IsIntegralConst(1))
```
There seems to be another opportunity for `IsIntegralConst` a few lines above, 7834-7837.

mikedn commented on src/jit/optimizer.cpp:
```diff
@@ -993,8 +993,7 @@ bool Compiler::optIsLoopTestEvalIntoTemp(GenTreePtr testStmt, GenTreePtr* newTes
         (opr1->OperGet() == GT_LCL_VAR) &&
-        (opr2->OperGet() == GT_CNS_INT) && 
-        (opr2->gtIntCon.gtIconVal == 0))
+        opr2->IsIntegralConst(0))
```
@sivarv left a comment about this before, it's not clear to me if it was addressed. Is allowing longs here valid? I suppose that it should be unless there are problems on x86.

CarolEidt replied:
There are checks later that the expressions involved are TYP_INT. I don't think it fundamentally should be necessary, but since it's there, I've re-added the check that it's GT_CNS_INT, so that we don't go further down the path before failing to match.

mikedn commented:
Nice!

The GitHub diff viewer throws a fit on gentree.h so I can't comment there. The FP 0 check was and still is likely wrong. `x == 0.0` is true even if `x` is `-0.0` and this means that a `-0.0` may end up being converted to `0.0` when codegen generates a `xor`. Of course, this isn't related to this PR and fixing this, if needed, should be done in another PR.

mikedn commented:
> The question I have is, Is native compiler clever enough to optimize out additional gtOper check done by IsIntOne()/IsIntMinuOne() or is it best to avoid it in the first place? 

I did some tests with VC++ and such redundant checks seem to be eliminated as expected. And the `GT_CNS_LNG` check is also discarded if previously oper was tested to be `GT_CNS_INT`. Of course, there may be cases where the compiler cannot prove that oper didn't change but such cases are rare and the perf impact should be minimal (such cases would most likely involve an interfering function call that the compiler doesn't have visibility into).

CarolEidt commented:
@mikedn - I've re-added gentree.h. For some reason I've had a couple of cases where my line endings got messed up (and I haven't even changed OSes on this one!). The diffs should be viewable now. Also, thanks for doing the analysis of the native compiler behavior!
I'll open an issue for the FPZero case.

CarolEidt commented:
@dotnet-bot test Windows_NT x86 legacy_backend Checked Build and Test

CarolEidt commented:
I discovered that I was missing some changes for arm64, so I added those. I also ran diffs - the optimizer.cpp changes were causing changes for x86. So, I narrowed it to `GT_CNS_INT` only.

sivarv commented on src/jit/morph.cpp:
```diff
@@ -12453,22 +12443,12 @@ GenTree* Compiler::fgMorphSmpOpOptional(GenTreeOp* tree)
-                        {
-                            ssize_t     icon = op2->gtOp.gtOp2->gtIntCon.gtIconVal;
-
-                            noway_assert(typ <= TYP_UINT);
```
This noway_assert got deleted. Looks like this optimization was done only for TYP_INT case on 32-bit but not for long/ulong.  With this change this optimization gets enabled for TYP_LONG/ULONG. Good thing though.  May be we need to modify this noway_assert to 

noway_assert(typ <= TYP_ULONG);

sivarv commented on src/jit/morph.cpp:
```diff
@@ -12739,20 +12713,13 @@ GenTree* Compiler::fgMorphSmpOpOptional(GenTreeOp* tree)
-            }
-            else if ((op2->gtOper == GT_CNS_INT) && (op2->gtIntConCommon.IconValue() == 1) &&
-                     op1->OperIsCompare())
+            else if (op2->IsIntegralConst(1) && op1->OperIsCompare())
```
Observation: Earlier we were performing this optimization only on TYP_INT on 32-bit and this change will perform on TYP_LONG as well.  Good change to extend to TYP_LONG.

CarolEidt replied:
Yes, this one I had noted, but also did not result in any diffs. I'm not certain why the xor of -1 is handled for both int and long, but not xor of 1.

sivarv commented:
Looks good.

CarolEidt commented:
@dotnet-bot test  Ubuntu x64 Checked Build and Test
